File: #0872.leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        
        vector<int> leaf1;
        vector<int> leaf2;
        leafValues(root1, &leaf1);
        leafValues(root2, &leaf2);
        return leaf1 == leaf2; 
    }

private:

    void leafValues(TreeNode* root, vector<int>* leaf) {

        stack<TreeNode*> stk;
        TreeNode* cur;
        
        stk.push(root);
        while (!stk.empty()) {
            cur = stk.top();
            stk.pop();
            
            if (cur->left == NULL && cur->right == NULL) {
                leaf->push_back(cur->val);
                continue;
            }

            if (cur->right) {
                stk.push(cur->right);
            }

            if (cur->left) {
                stk.push(cur->left);
            }
        }
    }
};

File: #0872.leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    let arr1 = [];
    dfs(root1, arr1);

    let arr2 = [];
    dfs(root2, arr2);
    return arr1.toString() === arr2.toString();
};

let dfs = (node, arr) => {
    if (!node) {
        return;
    }

    if (!node.left && !node.right) {
        arr.push(node.val)
        return
    }

    dfs(node.left, arr);
    dfs(node.right, arr);
};
